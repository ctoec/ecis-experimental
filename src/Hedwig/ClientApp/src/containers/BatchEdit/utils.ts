import { Enrollment } from "../../generated";
import { hasValidationErrors } from "../../utils/validations";

export const hasChildInfoSectionErrors = (enrollment: Enrollment) => 	hasValidationErrors(enrollment.child, [
		'birthdate',
		'birthCertificateId',
		'birthtown',
		'birthstate',
		'hispanicOrLatinxEthnicity',
		'americanIndianOrAlaskaNative',
		'asian',
		'blackOrAfricanAmerican',
		'nativeHawaiianOrPacificIslander',
		'white',
		'gender',
	]);

	export const hasFamilyInfoSectionErrors = (enrollment: Enrollment) => hasValidationErrors(enrollment.child?.family, undefined, true);

	export const hasFamilyIncomeSectionErrors = (enrollment: Enrollment) => hasValidationErrors(enrollment.child?.family, ['determinations']);

	export const hasEnrollmentFundingSectionErrors = (enrollment: Enrollment) => 
		hasValidationErrors(enrollment, ['entry']) ||
		hasValidationErrors(enrollment.child, ['c4KCertificateFamilyId', 'c4KCertificates']);


