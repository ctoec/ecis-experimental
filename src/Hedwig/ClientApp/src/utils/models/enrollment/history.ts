import { ProcessStepProps } from '../../../components/ProcessList/ProcessStep';
import { DeepNonUndefineable, DeepNonUndefineableArray } from '../../types';
import {
	Enrollment,
	Funding,
	FundingSource,
	FamilyDetermination,
	C4KCertificate,
} from '../../../generated';
import { prettyAge, prettyFundingTime, NO_FUNDING } from '..';
import dateFormatter from '../../dateFormatter';
import { dateSorter } from '../../dateSorter';
import moment from 'moment';
import { c4kCertificateSorter } from '../c4kCertificate';
import { fundingStartSorter, getFundingTime } from '../funding';

// sortable process step props include:
// - sortDate, the field to use for sorting
// - sortWeight, a number to use as a sorting tiebreaker to ensure enrollment starts come before funding/determination steps,
// 	and enrollment ends come after
type SortableProcessStepProps = ProcessStepProps & { sortDate: Date | null; sortWeight: number };

/**
 * Generates SortableProcessStepProps array for a given array of enrollments,
 * also creating step props for fundings and family determinations
 * @param enrollment
 * @returns sorted SortableProcessStepProps array
 */
export function getEnrollmentHistoryProps(
	currentEnrollment: DeepNonUndefineable<Enrollment>
): ProcessStepProps[] {
	// allEnrollments is sorted in descending order by Entry
	// (pastEnrollments are sorted thusly in API response)
	var allEnrollments = [
		currentEnrollment,
		...(currentEnrollment.pastEnrollments ? currentEnrollment.pastEnrollments : []),
	];

	var processStepProps: SortableProcessStepProps[] = [];

	allEnrollments.forEach((enrollment, idx) => {
		processStepProps = processStepProps.concat(
			enrollmentStepProps(enrollment, idx, allEnrollments.length)
		);

		// funding steps
		processStepProps = processStepProps.concat(
			fundingStepProps(enrollment.fundings, enrollment.exit)
		);

		// only add determination steps and C4K funding steps for the current enrollment (others will be dupes)
		if (idx === 0) {
			// determination steps
			processStepProps = processStepProps.concat(
				determinationStepProps(enrollment.child.family.determinations)
			);

			// C4K certification steps
			processStepProps = processStepProps.concat(
				c4kCertificateStepProps(enrollment.child.c4KCertificates)
			);
		}
	});

	// return all steps sorted by `sortDate`, and sortWeight
	return processStepProps.sort(enrollmentHistoryPropsSorter);
}

export function enrollmentStepProps(
	_enrollment: DeepNonUndefineable<Enrollment>,
	enrollmentIndex: number,
	allEnrollmentsCount: number
) {
	var processStepProps: SortableProcessStepProps[] = [];

	// enrollment start step
	if (_enrollment.entry) {
		processStepProps.push(
			getSortableStep({
				heading:
					enrollmentIndex === allEnrollmentsCount - 1
						? // heading text for first enrollment (last in ordered array)
						  `Enrolled in ${_enrollment.site.name}`
						: `Changed enrollment to ${prettyAge(_enrollment.ageGroup)} in ${
								_enrollment.site.name
						  }`,
				body: `on ${dateFormatter(_enrollment.entry, false)}`,
				stepDate: _enrollment.entry,
				stepWeight: -1,
			})
		);
	}

	// enrollment stop step (only care about withdrawal for the last enrollment)
	if (enrollmentIndex === 0 && _enrollment.exit) {
		processStepProps.push(
			getSortableStep({
				heading: `Withdrawn from ${_enrollment.site.name}`,
				body: `on ${dateFormatter(_enrollment.exit, false)}`,
				stepDate: _enrollment.exit,
				stepWeight: 1,
			})
		);
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of fundings
 * @param fundings
 * @param enrollmentExit the value of Exit on the owning enrollment,
 * used to determine if switch to private pay step should be created
 */
export function fundingStepProps(
	fundings: DeepNonUndefineable<Funding[]> | null,
	enrollmentExit: Date | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (fundings) {
		fundings.sort(fundingStartSorter).forEach(funding => {
			// CDC Funding start if:
			// - source is CDC
			// - funding has first reporting period
			if (funding.source === FundingSource.CDC && funding.firstReportingPeriod) {
				var heading = `Switched funding to ${funding.source}`;
				var fundingTime = getFundingTime(funding);
				if (fundingTime) heading += ` - ${prettyFundingTime(fundingTime)}`;

				processStepProps.push(
					getSortableStep({
						heading,
						body: `on ${dateFormatter(funding.firstReportingPeriod.periodStart, false)}`,
						stepDate: funding.firstReportingPeriod.periodStart,
					})
				);
			}

			// Switch to private pay if:
			// - source is CDC
			// - funding has last reporting period
			// - enrollment is still ongoing (exit == null)
			// 	 OR last reporting period ends before enrollment exit
			// TODO: Handle other funding sources
			if (
				funding.source === FundingSource.CDC &&
				funding.lastReportingPeriod &&
				(!enrollmentExit || funding.lastReportingPeriod.periodEnd < enrollmentExit)
			) {
				var dayAfterLastReportingPeriodEnd = moment(funding.lastReportingPeriod.periodEnd)
					.add(1, 'days')
					.toDate();
				processStepProps.push(
					getSortableStep({
						heading: `Switched funding to ${NO_FUNDING}`,
						body: `on ${dateFormatter(dayAfterLastReportingPeriodEnd, false)}`,
						stepDate: dayAfterLastReportingPeriodEnd
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of family determinations
 * @param determinations
 */
export function determinationStepProps(
	determinations: DeepNonUndefineable<FamilyDetermination[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];
	if (determinations) {
		determinations.forEach(determination => {
			if (!determination.notDisclosed && determination.determinationDate) {
				processStepProps.push(
					getSortableStep({
						heading: 'Income redetermined',
						body: `on ${dateFormatter(determination.determinationDate, false)}`,
						stepDate: determination.determinationDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of C4KCertificates
 * @param c4KCertificates
 */
export function c4kCertificateStepProps(
	c4KCertificates: DeepNonUndefineable<C4KCertificate[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (c4KCertificates) {
		c4KCertificates.sort(c4kCertificateSorter).forEach((c4kCertificate, idx) => {
			if (c4kCertificate.startDate) {
				processStepProps.push(
					getSortableStep({
						heading:
							idx === 0 ? 'Care 4 Kids certificate added' : 'Care 4 Kids certificate renewed',
						body: `on ${dateFormatter(c4kCertificate.startDate, false)}`,
						stepDate: c4kCertificate.startDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Sorts SortableProcessStepProp objects inversely by sortDate, using sortWeight as a tie breaker
 * @param a
 * @param b
 */
export function enrollmentHistoryPropsSorter(
	a: SortableProcessStepProps,
	b: SortableProcessStepProps
) {
	var dateSorterRes = dateSorter(a.sortDate, b.sortDate, true);

	if (dateSorterRes !== 0) return dateSorterRes;

	if (a.sortWeight < b.sortWeight) return 1;
	if (a.sortWeight > b.sortWeight) return -1;
	return 0;
}

/**
 * Returns a SortableProcessStepProps object composed from the given parameters,
 * using the stepDate as sortDate, and to get correct value for isNew
 * @param heading
 * @param body
 * @param stepDate
 * @param stepWeight
 */
export function getSortableStep(params: {
	heading: string;
	body: string;
	stepDate: Date;
	stepWeight?: number;
}): SortableProcessStepProps {
	return {
		heading: params.heading,
		body: params.body,
		sortDate: params.stepDate,
		sortWeight: params.stepWeight ? params.stepWeight : 0,
		isNew: stepIsNew(params.stepDate),
	};
}

/**
 * Returns true is date is less than 1 week old, else false
 * @param stepDate
 */
export function stepIsNew(stepDate: Date) {
	return moment().add(-1, 'week') < moment(stepDate);
}
