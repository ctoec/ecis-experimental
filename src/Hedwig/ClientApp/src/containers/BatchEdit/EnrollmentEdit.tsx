import React, { useContext } from 'react';
import UserContext from '../../contexts/User/UserContext';
import { ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest, Enrollment } from '../../generated';
import { getIdForUser } from '../../utils/models';
import useApi from '../../hooks/useApi';
import StepList, { StepProps } from '../../components/StepList/StepList';
import ChildInfo from './_sections/ChildInfo';
import { Section, SectionProps } from '../Enrollment/enrollmentTypes';
import { hasValidationErrors } from '../../utils/validations';

type EnrollmentEditProps = {
	enrollmentId: number;
	siteId: number;
}

export const EnrollmentEdit: React.FC<EnrollmentEditProps> = ({
	enrollmentId,
	siteId,
}) => {
	const { user } = useContext(UserContext);
	// get the enrollment detail
	const params: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest = {
		id: enrollmentId,
		siteId,
		orgId: getIdForUser(user, 'org'),
	}

	const { error, data: enrollment, loading } = useApi<Enrollment>(
		(api) => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(params),
		{ skip: !user }
	)


	if (!enrollment) {
		return <>NOT FOUND</>;
	}

	const props: SectionProps = {
		enrollment,
		siteId,
		
	}
	const sections: Section[] = [];
	if(hasValidationErrors(enrollment.child, undefined, true)))
	const steps = sections.map((section) => {
		const editPath = section.key;
		return { ...section, editPath } as StepProps<SectionProps>
	})
	return (
		<StepList steps={steps} props={} />
	)
}
