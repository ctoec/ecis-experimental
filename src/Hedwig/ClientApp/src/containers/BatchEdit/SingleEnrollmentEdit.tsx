import React, { useState, useContext, useEffect } from 'react';
import { useHistory } from 'react-router';
import { Enrollment } from '../../generated';
import { hasValidationErrors } from '../../utils/validations';
import { StepProps, Button, StepList } from '../../components';
import { lastFirstNameFormatter } from '../../utils/stringFormatters';
import dateFormatter from '../../utils/dateFormatter';
import UserContext from '../../contexts/User/UserContext';
import { getIdForUser, enrollmentWithDefaultFamily } from '../../utils/models';
import useApi from '../../hooks/useApi';
import { BatchEditStepProps } from './_sections/batchEditTypes';
import ChildInfo from './_sections/ChildInfo';
import FamilyInfo from './_sections/FamilyInfo';
import FamilyIncome from './_sections/FamilyIncome';
import EnrollmentFunding from './_sections/EnrollmentFunding';
import useCatchAllErrorAlert from '../../hooks/useCatchAllErrorAlert';

type SingleEnrollmentEditProps = {
	enrollmentId: number;
	updateEnrollments: (_: Enrollment) => void;
	siteId: number;
	moveNextEnrollment: () => void;
};

export const SingleEnrollmentEdit: React.FC<SingleEnrollmentEditProps> = ({
	enrollmentId,
	updateEnrollments,
	siteId,
	moveNextEnrollment,
}) => {
	const { user } = useContext(UserContext);
	const orgId = getIdForUser(user, 'org');

	const params = {
		orgId,
		siteId: siteId,
		id: enrollmentId,
	};
	const history = useHistory();
	const [mutatedEnrollment, setMutatedEnrollment] = useState<Enrollment>();

	// Update the list of enrollments in parent EnrollmentsEditList
	// to have the most current version of this single enrollment
	useEffect(() => {
		if (mutatedEnrollment) {
			updateEnrollments(mutatedEnrollment);
		}
	}, [mutatedEnrollment]);

	// populate mutatedEnrollment (better name: to-be-mutated-enrollment) with enrollment detail from backend
	const { data: enrollmentDetail } = useApi<Enrollment>(
		(api) => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(params),
		{
			successCallback: (returnedEnrollment) => {
				setMutatedEnrollment(
					returnedEnrollment.child?.family
						? returnedEnrollment
						: enrollmentWithDefaultFamily(returnedEnrollment, orgId)
				);
			},
			skip: !user,
			deps: [enrollmentId],
		}
	);

	// Create steps for step list, based on state of missing/needed information
	// in the fetched enrollment
	const steps: StepProps<BatchEditStepProps>[] = [];
	if (hasValidationErrors(enrollmentDetail?.child, undefined, true) /*skip subobj validation*/) {
		steps.push(ChildInfo);
	}
	if (hasValidationErrors(enrollmentDetail?.child?.family, undefined, true)) {
		steps.push(FamilyInfo);
	}
	if (hasValidationErrors(enrollmentDetail?.child?.family, ['determinations'])) {
		steps.push(FamilyIncome);
	}
	if (
		hasValidationErrors(enrollmentDetail, ['entry']) ||
		hasValidationErrors(enrollmentDetail?.child, ['c4KCertificateFamilyId', 'c4KCertificates'])
	) {
		steps.push(EnrollmentFunding);
	}

	// Set url path for initial step
	const firstStep = steps.length ? steps[0].key : 'complete';
	useEffect(() => {
		let path = '';
		const pathIdMatch = history.location.pathname.match(/(\d+)/);
		if (!pathIdMatch || pathIdMatch[0] !== `${enrollmentId}`) {
			path += `/batch-edit/${enrollmentId}`;
		}

		path += `#${firstStep}`;
		history.push(path);
	}, [enrollmentDetail]);

	// set up function to advance to next step.
	// If there is no next step for this enrollment,
	// then invoke the moveNextEnrollment function
	const locationHash = history.location.hash;
	const activeStepId = locationHash ? locationHash.slice(1) : firstStep;

	const moveNextStep = () => {
		const currentIndex = steps.findIndex((step) => step.key === activeStepId);
		if (currentIndex === steps.length - 1) {
			moveNextEnrollment();
			return;
		}

		history.push(`#${steps[currentIndex + 1].key}`);
	};

	// Set up PUT request, to be triggered by steplist forms
	const [attemptingSave, setAttemptingSave] = useState(false);
	const { error: errorOnSave } = useApi<Enrollment>(
		(api) =>
			api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdPut({
				...params,
				enrollment: mutatedEnrollment,
			}),
		{
			skip: !user || !attemptingSave,
			callback: () => setAttemptingSave(false),
			successCallback: (returnedEnrollment) => {
				setAttemptingSave(false);
				setMutatedEnrollment(returnedEnrollment);
				moveNextStep();
			},
		}
	);
	const errorAlertState = useCatchAllErrorAlert(errorOnSave);

	if (!mutatedEnrollment) {
		return <></>;
	}

	const stepProps: BatchEditStepProps = {
		enrollment: mutatedEnrollment,
		error: errorOnSave,
		errorAlertState,
		onSubmit: (userModifiedEnrollment) => {
			setMutatedEnrollment(userModifiedEnrollment);
			setAttemptingSave(true);
		},
		onSkip: moveNextStep,
	};

	return (
		<>
			<div className="padding-x-2 padding-bottom-3">
				<div className="flex-row display-flex flex-justify flex-align-end">
					<h2>{lastFirstNameFormatter(mutatedEnrollment.child)}</h2>
					<div className="text-baseline">
						Date of Birth: {dateFormatter(mutatedEnrollment.child?.birthdate)}
					</div>
				</div>
				<div className="margin-top-1">
					<Button
						appearance="unstyled"
						text="View full profile"
						href={`/roster/sites/${siteId}/enrollments/${enrollmentId}`}
					/>
				</div>
			</div>
			<div className="padding-top-1 border-top-1px border-base-light">
				{steps.length > 0 ? (
					<StepList
						key={mutatedEnrollment.id}
						steps={steps}
						props={stepProps}
						activeStep={activeStepId}
						type="embedded"
						headerLevel="h3"
					/>
				) : (
					<div className="margin-y-2 display-flex flex-center">All complete!</div>
				)}
			</div>
		</>
	);
};
