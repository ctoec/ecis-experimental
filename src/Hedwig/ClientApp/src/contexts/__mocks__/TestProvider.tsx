import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, History } from 'history';
import AuthenticationContext from '../Authentication/AuthenticationContext';
import UserContext from '../User/UserContext';
import ReportingPeriodContext from '../ReportingPeriod/ReportingPeriodContext';
import { User, ReportingPeriod } from '../../generated';
import { cdcReportingPeriods as _cdcReportingPeriods, user as _user } from '../../tests/data';

type TestProviderProps = {
	user?: User;
	cdcReportingPeriods?: ReportingPeriod[];
	history?: History<any>;
};

const TestProvider: React.FC<TestProviderProps> = ({
	children,
	user = _user,
	cdcReportingPeriods = _cdcReportingPeriods,
	history = createMemoryHistory(),
}) => {
	return (
		<AuthenticationContext.Provider
			value={{ accessToken: '', withFreshToken: jest.fn(), loading: false }}
		>
			<UserContext.Provider value={{ user, loading: false }}>
				<ReportingPeriodContext.Provider value={{ cdcReportingPeriods }}>
					<Router history={history}>{children}</Router>
				</ReportingPeriodContext.Provider>
			</UserContext.Provider>
		</AuthenticationContext.Provider>
	);
};

export default TestProvider;
