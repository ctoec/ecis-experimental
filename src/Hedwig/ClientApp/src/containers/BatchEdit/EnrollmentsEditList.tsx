import React, { useState } from 'react';
import { Enrollment } from '../../generated';
import { SideNav, TextWithIcon, InlineIcon } from '../../components';
import { lastFirstNameFormatter } from '../../utils/stringFormatters';
import { hasValidationErrors } from '../../utils/validations';
import { SingleEnrollmentEdit } from './SingleEnrollmentEdit';
import { ReactComponent as Success } from '../../../node_modules/uswds/dist/img/alerts/success.svg';
import { Link } from 'react-router-dom';
import { getMissingInfoPrettyString } from '../../utils/validations/getMissingInfoPrettyString';

type EnrollmentsEditListProps = {
	enrollments: Enrollment[];
	activeEnrollmentId?: number;
};

export const EnrollmentsEditList: React.FC<EnrollmentsEditListProps> = ({
	enrollments,
	activeEnrollmentId,
}) => {
	const [editedEnrollments, setEditedEnrollments] = useState(enrollments);

	// Function to enable SingleEnrollmentEdit components to update the
	// single enrollment they control in the list of edited enrollments.
	// Will be passed down into SingleEnrollmentEdit
	const replaceEnrollment = (updatedEnrollment: Enrollment) => {
		setEditedEnrollments((enrollments) => {
			const enrollmentIdx = enrollments.findIndex((e) => e.id === updatedEnrollment.id);
			enrollments[enrollmentIdx] = updatedEnrollment;
			return enrollments;
		});
	};

	const [currentEnrollmentIdx, setCurrentEnrollmentIdx] = useState<number | undefined>(
		activeEnrollmentId
			? editedEnrollments.findIndex((e) => e.id === activeEnrollmentId)
			: editedEnrollments.length
			? 0
			: undefined
	);

	const moveNext = () => {
		if (currentEnrollmentIdx === editedEnrollments.length - 1) {
			const stillMissingInfoEnrollmentIdx = editedEnrollments.findIndex((e) =>
				hasValidationErrors(e)
			);
			setCurrentEnrollmentIdx(
				stillMissingInfoEnrollmentIdx < 0 ? undefined : stillMissingInfoEnrollmentIdx
			);
			return;
		}

		setCurrentEnrollmentIdx((idx) => (idx !== undefined ? idx + 1 : idx));
	};

	return (
		<SideNav
			externalActiveItemIndex={currentEnrollmentIdx}
			items={editedEnrollments.map((enrollment) => ({
				title: hasValidationErrors(enrollment) ? (
					lastFirstNameFormatter(enrollment.child)
				) : (
					<TextWithIcon
						iconSide="right"
						text={lastFirstNameFormatter(enrollment.child)}
						Icon={Success}
						iconClassName="oec-inline-icon--complete"
					/>
				),
				description: getMissingInfoPrettyString(enrollment),
				content: (
					<SingleEnrollmentEdit
						key={enrollment.id}
						enrollment={enrollment}
						updateEnrollments={replaceEnrollment}
						siteId={enrollment.siteId}
						moveNextEnrollment={moveNext}
					/>
				),
			}))}
			noActiveItemContent={
				<div className="margin-x-4 margin-y-2 grid-row flex-align-center flex-column">
					<InlineIcon icon="complete" />
					<p className="text-bold">All children are up to date!</p>
					<Link to="/roster">Return to roster</Link>
				</div>
			}
		/>
	);
};
