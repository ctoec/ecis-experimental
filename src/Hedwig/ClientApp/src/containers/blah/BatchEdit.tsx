import React, { useContext, useState, useEffect } from 'react';
import UserContext from '../../contexts/User/UserContext';
import useApi from '../../hooks/useApi';
import { getIdForUser } from '../../utils/models';
import { Enrollment } from '../../generated';
import CommonContainer from '../CommonContainer';
import { Suspend } from '../../components/Suspend/Suspend';
import { SideNav, InlineIcon, Button, TextWithIcon, TextInput, DateRange } from '../../components';
import { lastFirstNameFormatter } from '../../utils/stringFormatters';
import { SideNavItemProps } from '../../components/SideNav/SideNavItem';
import { hasValidationErrors } from '../../utils/validations';
import { ReactComponent as ArrowRight } from '../../assets/images/arrowRight.svg';
import { EnrollmentEdit } from './SingleEnrollmentEdit';
import moment from 'moment';
import { BatchEditInternal } from './Internal';


type BatchEditProps = { };
export default function BatchEdit({
}: BatchEditProps) {
	const dateRange: DateRange = {startDate: moment(), endDate: moment() };
	const { user } = useContext(UserContext);

	const [refetch, setRefetch] = useState(0);
	const { data: enrollments, loading, error } = useApi<Enrollment[]>(
		(api) => 
			api.apiOrganizationsOrgIdEnrollmentsGet({
				orgId: getIdForUser(user, 'org'),
				include: ['child', 'fundings', 'sites'],
				startDate: dateRange.startDate?.toDate(),
				endDate: dateRange.endDate?.toDate(),
			}),
			{
				// deps: [refetch]
			}
	);

	return (
		<CommonContainer backHref="/roster" backText="Back to program roster">
			<div className="grid-container">
				<h1>Add needed information</h1>
				<Suspend waitFor={!loading} fallback={<div>Loading...</div>}>
					<p className="usa-intro">{enrollments?.length} enrollments have missing or incomplete infor required to submit reports</p>
					<BatchEditInternal setRefetch={setRefetch} enrollments={enrollments} />
				</Suspend>
			</div>
		</CommonContainer>
	)

}
