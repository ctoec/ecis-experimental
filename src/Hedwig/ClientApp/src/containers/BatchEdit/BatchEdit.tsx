import React, { useContext } from 'react';
import moment from 'moment';
import UserContext from '../../contexts/User/UserContext';
import useApi from '../../hooks/useApi';
import { Enrollment, FundingSource } from '../../generated';
import { getIdForUser, isFunded } from '../../utils/models';
import CommonContainer from '../CommonContainer';
import { EnrollmentsEditList } from './EnrollmentsEditList';

type BatchEditProps = {
	history: History;
	match: {
		params: { activeEnrollmentId?: number };
	};
};
const BatchEdit: React.FC<BatchEditProps> = ({ match: { params: sectionId } }) => {
	// TODO get from QS param from roster
	const startDate = moment();
	const endDate = moment();

	// Set up enrollments get.
	const { user } = useContext(UserContext);
	const params = {
		orgId: getIdForUser(user, 'org'),
		startDate: startDate.toDate(),
		endDate: endDate.toDate(),
	};

	const { data: enrollments, loading } = useApi<Enrollment[]>((api) =>
		api.apiOrganizationsOrgIdEnrollmentsGet(params)
	);

	if (loading) {
		return <>Loading...</>;
	}

	const needInfoEnrollments = (enrollments || []).filter(
		(e) => {
			console.log(e)
			return e.validationErrors && e.validationErrors.length && isFunded(e, { source: FundingSource.CDC })

		});
	return (
		<CommonContainer>
			<div className="grid-container">
				<h1>Add needed information</h1>
				<p className="usa-intro">
					{needInfoEnrollments.length} enrollments have missing or incomplete information
				</p>
				<EnrollmentsEditList enrollments={needInfoEnrollments} />
			</div>
		</CommonContainer>
	);
};

export default BatchEdit;
