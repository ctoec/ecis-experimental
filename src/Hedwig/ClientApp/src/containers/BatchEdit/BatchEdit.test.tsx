// Variables used in jest mocks -- must start with `mock`
import {
	mockReport,
	mockEnrollmentMissingBirthCertId, // full time preschool
	mockEnrollmentMissingAddress, // full time preschool,
} from '../../tests/data';
import mockUseApi, {
	mockApiOrganizationsOrgIdReportsIdGet,
} from '../../hooks/useApi/__mocks__/useApi';

let mockEnrollments = [mockEnrollmentMissingBirthCertId, mockEnrollmentMissingAddress];
// Jest mocks must occur before later imports
jest.mock('../../hooks/useApi', () => ({
	// When trying to mock both a default import and named import,
	// we must specify __esModule: true on the returned object.
	__esModule: true,
	default: mockUseApi({
		apiOrganizationsOrgIdReportsIdGet: (_: any) =>
			mockApiOrganizationsOrgIdReportsIdGet({ ...mockReport, enrollments: mockEnrollments })(_),
	}),
	paginate: (_: any, __: any) => _,
}));

let mockSiteId: number | undefined = 1;
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
	useParams: () => ({
		id: mockSiteId,
	}),
}));

import React from 'react';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';

import 'react-dates/initialize';

import { accessibilityTestHelper } from '../../tests/helpers';
import TestProvider from '../../contexts/__mocks__/TestProvider';

import BatchEdit from './BatchEdit';

const batchEditProps = {
	history: createMemoryHistory(),
	location: { search: 'TODO CHANGE THIS' },
};

describe('Batch edit', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(
			<TestProvider>
				<BatchEdit {...batchEditProps} />
			</TestProvider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	accessibilityTestHelper(
		<TestProvider>
			<BatchEdit {...batchEditProps} />
		</TestProvider>
	);
});
