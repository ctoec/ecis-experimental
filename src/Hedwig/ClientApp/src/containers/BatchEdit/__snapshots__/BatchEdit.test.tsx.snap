// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Batch edit matches snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
  />
  <div
    class="grid-container"
  >
    <h1>
      Add needed information
    </h1>
    <p
      class="usa-intro"
    >
      2 enrollments have missing or incomplete information
    </p>
    <div
      class="oec-sidenav grid-row"
    >
      <div
        class="mobile-lg:grid-col-4"
      >
        <nav>
          <ul>
            <li
              class="oec-sidenav__item active"
            >
              <button
                class="usa-button usa-button--unstyled"
                type="button"
              >
                <div>
                  <p
                    class="oec-sidenav-item__title"
                  >
                    Potter, Lily Luna 
                  </p>
                  <p
                    class="oec-sidenav-item__desc"
                  >
                    Birth certificate ID
                  </p>
                </div>
              </button>
            </li>
            <li
              class="oec-sidenav__item"
            >
              <button
                class="usa-button usa-button--unstyled"
                type="button"
              >
                <div>
                  <p
                    class="oec-sidenav-item__title"
                  >
                    Potter, Lily Luna 
                  </p>
                  <p
                    class="oec-sidenav-item__desc"
                  >
                    1 fields
                  </p>
                </div>
              </button>
            </li>
          </ul>
        </nav>
      </div>
      <div
        class="oec-sidenav__content mobile-lg:grid-col-8"
      >
        <div
          class="padding-x-2 padding-bottom-3"
        >
          <div
            class="flex-row display-flex flex-justify flex-align-end"
          >
            <h2>
              Potter, Lily Luna
            </h2>
            <div
              class="text-baseline"
            >
              Date of Birth: 12/12/2016
            </div>
          </div>
          <div
            class="margin-top-1"
          >
            <a
              href="/roster/sites/1/enrollments/2"
            >
              View full profile
            </a>
          </div>
        </div>
        <div
          class="padding-top-1 border-top-1px border-base-light"
        >
          <ol
            class="oec-step-list embedded"
          >
            <li
              class="oec-step-list__step oec-step-list__step--active embedded"
            >
              <div
                class="oec-step-list__step__content"
              >
                <h3
                  class="oec-step-list__step__title"
                >
                  Child's information
                </h3>
                <div
                  class="oec-step-list__step__form"
                >
                  <form
                    autocomplete="off"
                    class="usa-form"
                    novalidate=""
                  >
                    <h3>
                      Birth Certificate
                    </h3>
                    <fieldset
                      aria-required="true"
                      class="grid-gap grid-row usa-fieldset display-inline-block"
                      id="birth-certificate-fields"
                    >
                      <legend
                        id="fieldset-legend-birth-certificate-fields"
                      >
                        <span
                          class="usa-sr-only"
                        >
                          Birth certificate
                        </span>
                      </legend>
                      <div
                        class="grid-gap grid-row flex-align-start"
                      >
                        <div
                          class="mobile-lg:grid-col-12"
                        >
                          <div
                            class=" usa-form-group"
                          >
                            <label
                              class="usa-label"
                              for="birthCertificateId"
                            >
                              Birth certificate ID # 
                            </label>
                            <input
                              aria-required="true"
                              class="usa-input"
                              id="birthCertificateId"
                              type="text"
                              value=""
                            />
                          </div>
                        </div>
                        <div
                          class="mobile-lg:grid-col-8 display-inline-block"
                        >
                          <div
                            class=" usa-form-group"
                          >
                            <label
                              class="usa-label"
                              for="birthTown"
                            >
                              Town 
                            </label>
                            <input
                              aria-required="true"
                              class="usa-input"
                              id="birthTown"
                              type="text"
                              value="Hogsmeade"
                            />
                          </div>
                        </div>
                        <div
                          class="mobile-lg:grid-col-4 display-inline-block"
                        >
                          <div
                            class=" usa-form-group"
                          >
                            <label
                              class="usa-label"
                              for="birthState"
                            >
                              State 
                            </label>
                            <input
                              aria-required="true"
                              class="usa-input"
                              id="birthState"
                              type="text"
                              value="CT"
                            />
                          </div>
                        </div>
                      </div>
                    </fieldset>
                    <input
                      class="usa-button"
                      type="submit"
                      value="Save and next"
                    />
                    <button
                      class="usa-button usa-button--outline"
                      type="button"
                    >
                      Skip
                    </button>
                  </form>
                </div>
              </div>
              <div
                class="oec-step-list__step__actions"
              />
            </li>
          </ol>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
